### PR TITLE
Add per-image build matrix configuration (SOFTWARE-5904)

### PR DESCRIPTION
**Pull request by PPathu**
Added a python file that creates a json list of images and its corresponding parameters that get passed to the build job in the main workflow. This list is created using a new file called build-config.json that can be configured per image in its directory and if this file doesn't exist the script will use a default configurable file located in the parent directory of all the images to build the image.

The new workflow can build and push images by calling the action, where each individual image built can have a configurable file that is specific to it and contains unique parameters for that image.

NOTE: Most images don't have a build-config.json only sssd and s3-backup currently have the custom config file. The file needs to be added to the other image directories with the parameters that we want the specific image to build from. The default config file also needs to be changed based on what we believe the 'default' should be. I'll discuss with Brian L at some point to figure out whats best and add these files accordingly.

### CODE REVIEW

**Comment by brianhlin** [comment]
Hey @PPathu , I'll have more time for a full review later today but can you rebase on `opensciencegrid/main` to resolve the merge conflict? Remember to backup your branch in case the rebase goes poorly!

**Comment by brianhlin** [comment]
@PPathu it looks like something's breaking here https://github.com/opensciencegrid/images/actions/runs/10201800645/job/28226119465?pr=204#step:6:29

File: opensciencegrid/s3-backup/build-config.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+    "standard_build": true,
+    "repo_build": false,
+    "base_os": ["el7"],
+    "osg_series": ["3.6"],
+    "base_repo": ["development"]
+  }
```
Comment (brianhlin) [inline opensciencegrid/s3-backup/build-config.json:7]: This PR should be limited to adding the framework by which we can change the matrix. If we decide that we want to change the set of tags that we build for this image, it should go in a different PR

File: opensciencegrid/sssd/build-config.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+    "standard_build": true,
+    "repo_build": false,
+    "base_os": ["el9, el8"],
+    "osg_series": ["23, 3.6"],
+    "base_repo": ["development"]
+  }
```
Comment (brianhlin) [inline opensciencegrid/sssd/build-config.json:7]: Same comment about this config file being outside the scope of this PR

File: opensciencegrid/default-build-config.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+    "standard_build": true,
+    "repo_build": false,
+    "base_os": ["el9"],
+    "osg_series": ["23"],
+    "base_repo": ["development", "testing", "release"]
+  }
```
Comment (brianhlin) [inline opensciencegrid/default-build-config.json:7]: You're missing a newline here. I think this is something that VS Code just /does/. Maybe @jhiemstrawisc knows the configuration to set to force it to stop doing it

File: scripts/build-job-matrix.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    except (FileNotFoundError, json.JSONDecodeError):
+        if default_config is not None:
+            return default_config
+        else:
+            raise
+
+def main(image_dirs):
+    print("Image directories:", image_dirs)  
+
+    default_config_path = 'opensciencegrid/default-build-config.json'
```
Comment (brianhlin) [inline scripts/build-job-matrix.py:23]: Nitpick: this should be a top level constant of the library. Like inserting this between the last import and your first function definition:

```
DEFAULT_CONFIG_PATH = 'opensciencegrid/default-build-config.json'
```

File: scripts/build-job-matrix.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        if not os.path.isdir(image_dir):
+            sys.exit(f"Error: Image directory '{image_dir}' does not exist.")
+        build_config_path = os.path.join(image_dir, 'build-config.json')
+
+        config = load_config(build_config_path, default_config)
+
+        image_name = os.path.basename(image_dir)
+
+        base_os_list = config['base_os'][0].split(', ')
+        osg_series_list = config['osg_series'][0].split(', ')
```
Comment (brianhlin) [inline scripts/build-job-matrix.py:38]: It looks like you had to grab the first item in the list and split on the `,` because in the `s3-backup` and `sssd` configs, you were storing the "lists" as a single comma-delimited string as the first (and only) item in the lists. You could make these look like line 39 if you forced the config to be actual lists.

File: scripts/build-job-matrix.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+        for base_os, osg_series, base_repo in combinations:
+            configuration_string = f"{base_os}-{osg_series}-{base_repo}-{config['standard_build']}-{config['repo_build']}"
+            include_list.append({"name": image_name, "config": configuration_string})
+
+    output_path = os.path.join('scripts', 'grouped_output.json')
+    os.makedirs('scripts', exist_ok=True)
+    with open(output_path, 'w') as outfile:
+        json.dump({"include": include_list}, outfile, indent=4)
+    print(f"Generated {output_path} with grouped image configurations")
```
Comment (brianhlin) [inline scripts/build-job-matrix.py:55]: I would skip writing the constructed matrix out to disk and instead just have the script dump it to stdout. The caller can easily decide if they want to write it to disk.

**Review by brianhlin — CHANGES_REQUESTED** [review]
A quick review of the Python script

File: opensciencegrid/default-build-config.json (hunk 3 above)
Comment (jhiemstrawisc) [reply opensciencegrid/default-build-config.json:7]: I do! If you go to your settings explorer (on Mac, that's `Code-->Settings-->Settings`) you can search for "Insert Final Newline". Check the box to save the setting and you should be good to go.

File: scripts/build-job-matrix.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        base_repo_list = config['base_repo']
+
+        combinations = product(
+            base_os_list,
+            osg_series_list,
+            base_repo_list
+        )
+
+        for base_os, osg_series, base_repo in combinations:
+            configuration_string = f"{base_os}-{osg_series}-{base_repo}-{config['standard_build']}-{config['repo_build']}"
```
Comment (brianhlin) [inline scripts/build-job-matrix.py:48]: At first glance, this is a really strange way to pass this information down to the consumer: this would benefit from a comment as to why you've done it this way.

File: .github/workflows/build-containers.yml [hunk 8]
```diff
@@ -111,80 +152,102 @@ jobs:
...
     steps:
-      - name: Push to Harbor (${{ matrix.series }}-development)
-        # FIXME: reenable for OSG 23 development when available
-        if: matrix.series.name != '23'
+      - uses: actions/checkout@v3
+
+      - name: Print raw matrix configuration
+        run: |
+          echo "Raw matrix configuration: ${{ matrix }}"
+      - name: Set environment variables
```
Comment (brianhlin) [inline .github/workflows/build-containers.yml:162]: This is hard to follow because it's not clear what the input format is (presumably the long config string?). A comment here with an example item from the JSON would be really helpful

**Review by brianhlin — APPROVED** [review]
LGTM for a first pass

File: scripts/build-job-matrix.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+            base_repo_list
+        )
+
+        for base_os, osg_series, base_repo in combinations:
+        # Constructs a unique configuration string to represent each build setting.
+        # This combines the base OS, OSG series, base repository, and two standard
+        # build parameters, making it easy to identify and manage each configuration.
+        # This approach is preferred because it offers:
+        # 1. Simplicity: Using a single string to represent configurations is straightforward and easy to understand.
+        # 2. Integration: A single string is easily passed to external tools and systems that manage builds.
```
[the remark this answers is not in the document]
Comment (PPathu) [reply scripts/build-job-matrix.py:54]: added this important detail in the next PR
